src/build/bundle.ts

```diff
@@ -9,55 +9,328 @@
  */
 
 import * as gulpif from 'gulp-if';
+import * as path from 'path';
 import {Transform} from 'stream';
+import File = require('vinyl');
 
 import {Logger} from './logger';
 import {compose} from './streams';
 import {StreamResolver} from './stream-resolver';
 import {VulcanizeTransform} from './vulcanize';
+import {Analyzer, Loader} from 'hydrolysis';
 
 // non-ES module
 const minimatchAll = require('minimatch-all');
+const through = require('through2').obj;
+const Vulcanize = require('vulcanize');
 
 export class Bundler {
 
   root: string;
   shell: string;
+  sharedBundleUrl: string;
   allEntrypoints: string[];
 
   streamResolver: StreamResolver;
+  loader: Loader;
   bundle: Transform;
   vulcanize: Transform;
+  analyzer: Analyzer;
+  entrypointFiles: Map<string, File>;
+  sharedFile: File;
 
-  constructor(root: string, shell: string, entrypoints: string[]) {
+  _verboseLogging = false;
+
+  constructor(root: string, shell: string, entrypoints?: string[]) {
     this.root = root;
     this.shell = shell;
-    let allEntrypoints = [];
-    if (shell) allEntrypoints.push(shell);
+    this.sharedBundleUrl = path.resolve(root, 'shared-bundle.html');
+
+    let _allEntrypoints = [];
+    // It's important that shell is first for document-ordering of imports
+    if (shell) _allEntrypoints.push(shell);
     this.allEntrypoints =
-        Array.prototype.concat.apply(allEntrypoints, entrypoints);
+        Array.prototype.concat.apply(_allEntrypoints, entrypoints);
 
     this.streamResolver = new StreamResolver({
         entrypoints: this.allEntrypoints,
         basePath: root,
         root: root,
         redirect: 'bower_components/',
       });
+
+    this.loader = new Loader();
+    this.loader.addResolver(this.streamResolver);
+    this.analyzer = new Analyzer(false, this.loader);
+
     this.vulcanize = new VulcanizeTransform({
       fsResolver: this.streamResolver,
       inlineScripts: true,
       inlineCss: true,
       stripComments: true,
     });
 
+    this.entrypointFiles = new Map();
+
     this.bundle = compose([
       this.streamResolver,
-      gulpif((file) => this.isEntryPoint(file.path), this.vulcanize)
+      through((file, enc, cb) => {
+        if (this.isEntryPoint(file.path)) {
+          // don't pass on any entrypoints until the stream has ended
+          this.entrypointFiles.set(file.path, file);
+          cb(null, null);
```

Comment:
I'm confident it doesn't, given that the entrypoints are eventually written out. This is covered in the tests.